pandas/tseries/offsets.py

```diff
@@ -221,6 +232,60 @@ def apply(self, other):
         else:
             return other + timedelta(self.n)
 
+    @apply_index_wraps
+    def apply_index(self, i):
+        """
+        Vectorized apply of DateOffset to DatetimeIndex,
+        raises NotImplentedError for offsets without a
+        vectorized implementation
+
+        Parameters
+        ----------
+        i : DatetimeIndex
+
+        Returns
+        -------
+        y : DatetimeIndex
+        """
+
+        relativedelta_fast = set(['years', 'months', 'weeks',
+                                  'days', 'hours', 'minutes',
+                                  'seconds', 'microseconds'])
+        # relativedelta/_offset path only valid for base DateOffset
+        if type(self) is DateOffset:
```

Comment:
The intent was to only execute this block for the base `DateOffset` class - it produces invalid deltas for subclasses that don't override `__init__` (e.g. `BusinessMonthBegin`).  Would it be better to explicitly raise `NotImplementedError` from those subclasses?